nixos/lib/systemd-lib.nix

```diff
@@ -235,4 +236,205 @@ in rec {
       ''}
     ''; # */
 
+  makeJobScript = name: text:
+    let
+      scriptName = replaceChars [ "\\" "@" ] [ "-" "_" ] (shellEscape name);
+      out = (pkgs.writeShellScriptBin scriptName ''
+        set -e
+        ${text}
+      '').overrideAttrs (_: {
+        # The derivation name is different from the script file name
+        # to keep the script file name short to avoid cluttering logs.
+        name = "unit-script-${scriptName}";
+      });
+    in "${out}/bin/${scriptName}";
+
+  unitConfig = { config, options, ... }: {
```

Comment:
I fully agree with your comments. However, this PR was made to have make it easier to merge. It is the minimal amount of changes that I need to reuse systemd options for systemd in initramfs.

This PR was split off from https://github.com/NixOS/nixpkgs/pull/164016. It was suggested to me to keep the PRs as small as possible to avoid long review a long review process and avoid it going out-of-date quickly. I think that is indeed the right way to go, so I'm going to keep creating smallish PRs to move towards the goal of having a cleaner systemd module as well as reusing parts in systemd.user and initramfs.

For your interest, I have also tried to create individual 'submodules' for each part of the common systemd options. `units`, `services`, `paths`, `mounts`, etc. That work is already done, but I'm going to wait to create a PR for that because I need as least the changes in this PR to accomplish it.

See https://github.com/bobvanderlinden/nixpkgs/commit/e5ad5b33604e9cb7bf2c37df179518140db7e430. I'm also very interested in feedback on this one... though I do know it will be a long while before that can even be considered for reviewing.

Reply:
> See [bobvanderlinden@e5ad5b3](https://github.com/bobvanderlinden/nixpkgs/commit/e5ad5b33604e9cb7bf2c37df179518140db7e430). I'm also very interested in feedback on this one...

I'd try to avoid functions to modules; instead turning the parameters into internal options.
Instead of `${unitType} =`, I'd try to make a generic common module and import it in each "normal" unit type submodule.
I guess this isn't as commonplace as it should be because of https://github.com/NixOS/nixpkgs/pull/156533

Reply:
The internal options would be ideal. I started out with `systemd = mkOption { type = submoduleWith { modules = [ units ... ]; } }`, but that resulted in problems. Iteratively I came to the solution in the branch. Can you explain how https://github.com/NixOS/nixpkgs/pull/156533 can make the branch better?